File: julia/Day8.jl
module Day8

export Submarine, on_input!, output

# This maps to (1 + digit) because julia's indices start at 1.
const EASY_LEN_TO_DIGIT = Dict(2=>2, 3=>8, 4=>5, 7=>9)
const EASY_DIGITS = [1, 7, 4, 8]

mutable struct Submarine
    total_easy_digits::Int32
    sum_digits::Int32
    digits::Vector{Set{Char}}
    five_len_possibilities::Vector{Set{Char}}
    six_len_possibilities::Vector{Set{Char}}

    Submarine() = new(0, 0, [Set() for _ in 1:10], [], [])
end

function on_input!(s::Submarine, a::AbstractString)
    parts = collect(split(a, " | "))

    foreach(x -> add_unique_set!(s, x), split(parts[1], " "))

    figure_out_ambiguous_digits!(s)

    display_digits = map(x -> digit_from_set(s, x), Set.(split(parts[2], " ")))

    s.total_easy_digits += length(filter(x -> x in EASY_DIGITS, display_digits))

    display_number = 1000 * display_digits[1] + 100 * display_digits[2] + 10 * display_digits[3] + display_digits[4]
    s.sum_digits += display_number
end

function add_unique_set!(s::Submarine, a::AbstractString)
    a_len = length(a)

    if a_len in keys(EASY_LEN_TO_DIGIT)
        s.digits[EASY_LEN_TO_DIGIT[a_len]] = Set(a)
    elseif a_len == 5
        push!(s.five_len_possibilities, Set(a))
    elseif a_len == 6
        push!(s.six_len_possibilities, Set(a))
    end
end

function removeelement!(s, e)
    deleteat!(s, findfirst(x -> x == e, s))
end

function figure_out_ambiguous_digits!(s::Submarine)
    # The set that represents 3 is the 5-digit that contains 1.
    three = first(filter(x -> s.digits[2] ⊆ x, s.five_len_possibilities))
    s.digits[4] = three
    removeelement!(s.five_len_possibilities, three)

    # The set that represents 5 is the 5-digit except 3 that has 3 intersections with 4.
    five = first(filter(x -> length(intersect(s.digits[5], x)) == 3, s.five_len_possibilities))
    s.digits[6] = five
    removeelement!(s.five_len_possibilities, five)

    # The set that represents 2 is the remaining 5-digit.
    s.digits[3] = pop!(s.five_len_possibilities)

    # The set that represents 0 is the 6-digit that doesn't contain 5.
    zero = first(filter(x -> s.digits[6] ⊈ x, s.six_len_possibilities))
    s.digits[1] = zero
    removeelement!(s.six_len_possibilities, zero)

    # The set that represents 9 is the 6-digit that contains 3.
    nine = first(filter(x -> s.digits[4] ⊆ x, s.six_len_possibilities))
    s.digits[10] = nine
    removeelement!(s.six_len_possibilities, nine)

    # The set that represents 6 is the remaining 6-digit.
    s.digits[7] = pop!(s.six_len_possibilities)
end

function digit_from_set(s::Submarine, x)
    findfirst(y -> issetequal(x, y), s.digits) - 1
end

function print_digits(s::Submarine)
    for (i, set) in enumerate(s.digits)
        println("$(i - 1) -> $(set)")
    end
end

function output(s::Submarine)
    println("Part 1: $(s.total_easy_digits)")
    println("Part 2: $(s.sum_digits)")
end

end

File: 2021/rust/src/day8.rs
use std::collections::HashSet;

const EASY_DIGITS: [usize; 4] = [1, 7, 4, 8];
const EASY_LENGTHS: [usize; 4] = [2, 3, 4, 7];

fn easy_length_to_digit(len: usize) -> usize {
    match len {
        2 => 1,
        3 => 7,
        4 => 4,
        7 => 8,
        _ => usize::MAX,
    }
}

pub struct Submarine {
    total_easy_digits: usize,
    sum_digits: usize,
    digits: [HashSet<char>; 10],
    five_len_possibilities: Vec<HashSet<char>>,
    six_len_possibilities: Vec<HashSet<char>>,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            total_easy_digits: 0,
            sum_digits: 0,
            digits: Default::default(),
            five_len_possibilities: Vec::new(),
            six_len_possibilities: Vec::new(),
        }
    }

    pub fn on_input(&mut self, result: &str) {
        let mut parts = result.split(" | ");

        // Storing the 10 unique sets for processing.
        parts.next().unwrap().split(" ").for_each(|set| self.add_unique_set(set));

        self.figure_out_ambiguous_digits();

        // Match each set from the display to the actual digit it represents.
        let display_digits: Vec<_> = parts.next().unwrap().split(" ").map(hashset_from_str).map(|s| self.digits.iter().position(|x| *x == s).unwrap()).collect();

        self.total_easy_digits += display_digits.iter().filter(|x| EASY_DIGITS.contains(x)).count();

        let display_number = 1000 * display_digits[0] + 100 * display_digits[1] + 10 * display_digits[2] + display_digits[3];
        self.sum_digits += display_number;
    }

    fn add_unique_set<'a>(&mut self, set: &'a str) {
        match set.len() {
            n if EASY_LENGTHS.contains(&n) => self.digits[easy_length_to_digit(n)] = hashset_from_str(set),
            5 => self.five_len_possibilities.push(hashset_from_str(set)),
            6 => self.six_len_possibilities.push(hashset_from_str(set)),
            _ => panic!("Got a set with weird length!"),
        }
    }

    fn figure_out_ambiguous_digits(&mut self) {
        // The set that represents 3 is the 5-digit that contains 1.
        let three_index = self.five_len_possibilities.iter().position(|x| x.is_superset(&self.digits[1])).unwrap();
        self.digits[3] = self.five_len_possibilities.swap_remove(three_index);

        // The set that represents 5 is the 5-digit except 3 that has 3 intersections with 4.
        let five_index = self.five_len_possibilities.iter().position(|x| x.intersection(&self.digits[4]).count() == 3).unwrap();
        self.digits[5] = self.five_len_possibilities.swap_remove(five_index);

        // The set that represents 2 is the remaining 5-digit.
        self.digits[2] = self.five_len_possibilities.pop().unwrap();

        // The set that represents 0 is the 6-digit that doesn't contain 5.
        let zero_index = self.six_len_possibilities.iter().position(|x| !x.is_superset(&self.digits[5])).unwrap();
        self.digits[0] = self.six_len_possibilities.swap_remove(zero_index);

        // The set that represents 9 is the 6-digit that contains 3.
        let nine_index = self.six_len_possibilities.iter().position(|x| x.is_superset(&self.digits[3])).unwrap();
        self.digits[9] = self.six_len_possibilities.swap_remove(nine_index);

        // The set that represents 6 is the remaining 6-digit.
        self.digits[6] = self.six_len_possibilities.pop().unwrap();
    }

    fn _print_digits(&self) {
        for (i, set) in self.digits.iter().enumerate() {
            let charset: String = set.iter().to_owned().collect();
            println!("{} -> {}", i, charset);
        }
    }

    pub fn output(&mut self) {
        println!("Part 1: {}", self.total_easy_digits);
        println!("Part 2: {}", self.sum_digits);
    }
}

fn hashset_from_str<'a>(set: &'a str) -> HashSet<char> {
    let mut hashset = HashSet::new();
    set.chars().for_each(|x| {
        hashset.insert(x);
    });

    hashset
}